File: frontend/src/services/sitio2/Mortalidad.ts
// src/services/sitio2/Mortalidad.ts
import API_BASE from "../../config/api";

export type RegistroMortalidad = {
  id: number;
  fecha: string;
  lote: string;
  corral: string;
  cantidad: number;
  causa: string;
  tipo: "Mortalidad" | "Descarte";
  responsable: string;
  observaciones: string;
};

export type NuevoRegistroMortalidad = Omit<RegistroMortalidad, "id">;

const EMPRESA_ID = 1;
const GRANJA_ID = 1;

function mapApiToMortalidad(api: any): RegistroMortalidad {
  return {
    id: api.id,
    fecha: api.fecha,
    lote: api.lote,
    corral: api.corral,
    cantidad: api.cantidad,
    causa: api.causa,
    tipo: api.tipo,
    responsable: api.responsable,
    observaciones: api.observaciones ?? "",
  };
}

export async function getMortalidadSitio2(): Promise<RegistroMortalidad[]> {
  const res = await fetch(
    `${API_BASE}/sitio2/mortalidad/?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`
  );
  if (!res.ok) {
    throw new Error("Error al obtener mortalidad y descartes");
  }
  const data = await res.json();
  return data.map(mapApiToMortalidad);
}

export async function addMortalidadSitio2(
  reg: NuevoRegistroMortalidad
): Promise<RegistroMortalidad> {
  const payload = {
    empresa_id: EMPRESA_ID,
    granja_id: GRANJA_ID,
    ...reg,
  };

  const res = await fetch(`${API_BASE}/sitio2/mortalidad/`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    throw new Error("Error al registrar baja/descarte");
  }
  const data = await res.json();
  return mapApiToMortalidad(data);
}

export async function updateMortalidadSitio2(
  id: number,
  reg: NuevoRegistroMortalidad
): Promise<RegistroMortalidad> {
  const res = await fetch(
    `${API_BASE}/sitio2/mortalidad/${id}?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`,
    {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(reg),
    }
  );
  if (!res.ok) {
    throw new Error("Error al actualizar baja/descarte");
  }
  const data = await res.json();
  return mapApiToMortalidad(data);
}

export async function deleteMortalidadSitio2(id: number): Promise<void> {
  const res = await fetch(
    `${API_BASE}/sitio2/mortalidad/${id}?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`,
    { method: "DELETE" }
  );
  if (!res.ok) {
    throw new Error("Error al eliminar baja/descarte");
  }
}

File: backend/api/sitio2/Mortalidad.py
# backend/api/sitio2/Mortalidad.py
from typing import List

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from backend.database import get_db
from backend.models.sitio2.Mortalidad import (
    MortalidadRepository,
    MortalidadRead,
    MortalidadCreate,
    MortalidadUpdate,
)
from backend.validators.crecimiento_validators import (
    CrecimientoValidators,
    DatosMortalidadSitio,
    CrecimientoValidationError,
)

router = APIRouter(
    prefix="/sitio2/mortalidad",
    tags=["Sitio 2 - Mortalidad y Descartes"],
)


def get_repo(db: Session = Depends(get_db)) -> MortalidadRepository:
    return MortalidadRepository(db)


@router.get(
    "/",
    response_model=List[MortalidadRead],
    summary="Listar mortalidad y descartes",
)
def listar_mortalidad(
    empresa_id: int,
    granja_id: int,
    repo: MortalidadRepository = Depends(get_repo),
):
    return repo.listar_por_granja(empresa_id, granja_id)


@router.post(
    "/",
    response_model=MortalidadRead,
    status_code=status.HTTP_201_CREATED,
    summary="Registrar baja/descarte",
)
def crear_mortalidad(
    payload: MortalidadCreate,
    repo: MortalidadRepository = Depends(get_repo),
    db: Session = Depends(get_db),
):
    try:
        CrecimientoValidators.validar_mortalidad(
            db=db,
            datos=DatosMortalidadSitio(
                empresa_id=payload.empresa_id,
                granja_id=payload.granja_id,
                lote=payload.lote,
                corral=payload.corral,
                cantidad=payload.cantidad,
                tipo=payload.tipo,
            ),
        )
    except CrecimientoValidationError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(exc),
        )

    return repo.crear(payload)


@router.put(
    "/{reg_id}",
    response_model=MortalidadRead,
    summary="Actualizar baja/descarte",
)
def actualizar_mortalidad(
    reg_id: int,
    empresa_id: int,
    granja_id: int,
    cambios: MortalidadUpdate,
    repo: MortalidadRepository = Depends(get_repo),
    db: Session = Depends(get_db),
):
    reg = repo.obtener_por_id(reg_id, empresa_id, granja_id)
    if not reg:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Registro de mortalidad/descarte no encontrado",
        )

    # Validar solo si se tocan datos críticos
    if any(
        getattr(cambios, campo) is not None
        for campo in ("lote", "corral", "cantidad", "tipo")
    ):
        datos = DatosMortalidadSitio(
            empresa_id=reg.empresa_id,
            granja_id=reg.granja_id,
            lote=cambios.lote or reg.lote,
            corral=cambios.corral or reg.corral,
            cantidad=(
                cambios.cantidad if cambios.cantidad is not None else reg.cantidad
            ),
            tipo=cambios.tipo or reg.tipo,
        )
        try:
            CrecimientoValidators.validar_mortalidad(db=db, datos=datos)
        except CrecimientoValidationError as exc:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=str(exc),
            )

    return repo.actualizar(reg, cambios)


@router.delete(
    "/{reg_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Eliminar baja/descarte",
)
def eliminar_mortalidad(
    reg_id: int,
    empresa_id: int,
    granja_id: int,
    repo: MortalidadRepository = Depends(get_repo),
):
    reg = repo.obtener_por_id(reg_id, empresa_id, granja_id)
    if not reg:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Registro de mortalidad/descarte no encontrado",
        )
    repo.eliminar(reg)
